File: src/services/cryptoService.js
// Cifra/descifra los valores "secretos" guardados en la tabla `settings`
// (API keys, tokens, contraseña SMTP) con AES-256-GCM, para que un dump
// de la base de datos no los exponga en texto plano. La clave sale de
// CREDENTIALS_ENC_KEY (64 caracteres hex = 32 bytes) en .env.
//
// Si esa variable no esta configurada, encrypt()/decrypt() se comportan
// como una funcion identidad (devuelven el valor tal cual) en vez de
// reventar - asi una instalacion ya en produccion sigue funcionando
// igual despues de un `git pull` aunque todavia no le hayan agregado la
// clave nueva al .env; en cuanto la agreguen y reinicien, empieza a
// cifrar solo.
const crypto = require('crypto');
const env = require('../config/env');

const ALGORITHM = 'aes-256-gcm';
const PREFIX = 'enc:v1:';

function getKey() {
  const hex = env.credentialsEncKey;
  if (!hex || hex.length !== 64) return null;
  return Buffer.from(hex, 'hex');
}

function isEncrypted(value) {
  return typeof value === 'string' && value.startsWith(PREFIX);
}

function encrypt(text) {
  const key = getKey();
  if (!key || text === undefined || text === null || text === '') return text;
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);
  const ciphertext = Buffer.concat([cipher.update(String(text), 'utf8'), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return `${PREFIX}${iv.toString('hex')}:${authTag.toString('hex')}:${ciphertext.toString('hex')}`;
}

function decrypt(value) {
  if (!isEncrypted(value)) return value; // legado en texto plano, o vacio
  const key = getKey();
  if (!key) return value; // sin clave configurada: no se puede descifrar

  const rest = value.slice(PREFIX.length); // "<ivHex>:<authTagHex>:<cipherHex>"
  const [ivHex, tagHex, dataHex] = rest.split(':');
  if (!ivHex || !tagHex || dataHex === undefined) return value; // formato inesperado

  try {
    const decipher = crypto.createDecipheriv(ALGORITHM, key, Buffer.from(ivHex, 'hex'));
    decipher.setAuthTag(Buffer.from(tagHex, 'hex'));
    const plain = Buffer.concat([decipher.update(Buffer.from(dataHex, 'hex')), decipher.final()]);
    return plain.toString('utf8');
  } catch (err) {
    // Clave rotada/incorrecta o dato corrupto - mejor devolver el valor
    // cifrado que reventar toda la app por un solo campo ilegible.
    return value;
  }
}

module.exports = { encrypt, decrypt, isEncrypted };

File: devops-sidecar/app/services/crypto_service.py
"""Cifra/descifra los valores "secretos" guardados en app_settings (API
keys de Gemini/Claude) con Fernet (AES-128-CBC + HMAC, autenticado), para
que un dump de la base de datos SQLite no los exponga en texto plano. La
clave sale de CREDENTIALS_ENC_KEY en .env (formato urlsafe-base64 de 32
bytes - el mismo que genera Fernet.generate_key(), o
`openssl rand 32 | base64 | tr '+/' '-_'`).

Si esa variable no esta configurada, encrypt()/decrypt() se comportan
como una funcion identidad (devuelven el valor tal cual) en vez de
reventar - misma logica que el lado Node (ver
glpi-licencias-app/src/services/cryptoService.js) - para que una
instancia ya en marcha siga funcionando igual hasta que se le agregue la
clave nueva y se reinicie el contenedor.
"""
from cryptography.fernet import Fernet, InvalidToken

from ..config import settings

PREFIX = "enc:v1:"


def _get_fernet() -> Fernet | None:
    key = settings.credentials_enc_key
    if not key:
        return None
    try:
        return Fernet(key.encode())
    except (ValueError, TypeError):
        return None


def is_encrypted(value) -> bool:
    return isinstance(value, str) and value.startswith(PREFIX)


def encrypt(text):
    fernet = _get_fernet()
    if not fernet or not text:
        return text
    token = fernet.encrypt(text.encode()).decode()
    return f"{PREFIX}{token}"


def decrypt(value):
    if not is_encrypted(value):
        return value  # legado en texto plano, o vacio
    fernet = _get_fernet()
    if not fernet:
        return value  # sin clave configurada: no se puede descifrar
    token = value[len(PREFIX):]
    try:
        return fernet.decrypt(token.encode()).decode()
    except (InvalidToken, ValueError):
        # Clave rotada/incorrecta o dato corrupto - mejor devolver el
        # valor cifrado que reventar la app por un solo campo ilegible.
        return value
